lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/BinarySpec.hs

```diff
@@ -203,3 +237,123 @@ spec = do
+        Transaction stx -> stx
+        _ -> error "expected a Transaction message"
+
+    try' :: a -> Either String a
+    try' = unsafePerformIO . wrap
+
+    wrap :: a -> IO (Either String a)
+    wrap = fmap (either (Left . show) Right) . (try @SomeException) . evaluate
+
+    labelTx stx = label (either id (const "Right _") (validate stx))
+
+    validate :: (Tx, [TxWitness]) -> Either String (Tx, [TxWitness])
+    validate stx@(Tx ins outs, _wits) = const stx <$> do
+        unless (length ins < 256) $
+            Left "number of inputs must be lower than 256"
+        unless (length outs < 256) $
+            Left "number of outputs must be lower than 256"
+        forM_ ins $ \(TxIn _tx ix, _coin) ->
+            unless (ix < 256) $ Left "input index must be < 256"
+
+
+-- Only generating single addresses!
+instance Arbitrary Address where
+    arbitrary = Address . appendTag 3 <$> genFixed 32
+    shrink (Address addr) = Address . appendTag 3 <$> shrinkFixedBS (BS.tail addr)
+
+-- Observation:
+-- genFixed and shrinkFixed would be nice candidates for DerivingVia.
+-- e.g.
+-- deriving instance Arbitrary Address via (ByteStringOfLength @33)
+genFixed :: Int -> Gen BS.ByteString
+genFixed n = BS.pack <$> (vectorOf n arbitrary)
+
+shrinkFixedBS :: ByteString -> [ByteString]
+shrinkFixedBS bs = if bs /= zeros then [zeros] else []
+      where
+        len = BS.length bs
+        zeros = BS.pack (replicate len 0)
+
+instance Arbitrary (Hash "Tx") where
+    arbitrary = Hash <$> genFixed 32
+    shrink (Hash bytes) = Hash <$> shrinkFixedBS bytes
+
+instance Arbitrary Coin where
+    arbitrary = genericArbitrary
+    shrink = genericShrink
+
+-- | Generates mostly valid input indexes, but a few invalid ones.
+instance Arbitrary TxIn where
+    arbitrary = TxIn
+        <$> arbitrary
+        <*> (frequency
+            [ (99, fromIntegral <$> arbitrary @Word8)
+            , (1, arbitrary @Word32)
+            ])
+    shrink = genericShrink
+
+instance Arbitrary TxOut where
+    arbitrary = genericArbitrary
+    shrink = genericShrink
+
+instance Arbitrary Tx where
+    -- We use scale (*6) to get large enough txs to hit the 256 limit
+    -- on number of inputs and outputs.
+    arbitrary = scale (*6) genericArbitrary
+
+newtype SignedTx = SignedTx (Tx, [TxWitness])
+    deriving (Eq, Show, Generic)
+
+instance Arbitrary SignedTx where
+    arbitrary = do
+        tx <- arbitrary
+        witnesses <- vectorOf (length $ inputs tx) arbitrary
+        return $ SignedTx (tx, witnesses)
+    shrink (SignedTx (tx, wits)) =
+         let
+            inps = inputs tx
+            zs = zip inps wits
+         in do
+            -- Try fewer inputs and outputs
+            --
+            -- NOTE: We don't shrink the individual inputs/outputs/witnesses.
+            -- Some of the shrinkers in this file are therefore unused.
+            xs <- take 5 $ shrinkList return zs
+            outs' <- take 5 $ shrinkList return $ outputs tx
```

Comment:
```hs
shrinkList :: (a -> [a]) -> [a] -> [[a]] 
```

So, `shrinkList return` would yield the value itself, which will cause QuickCheck to loop endlessly. It feels also a bit weird to also `take 5` elements from the shrinked list ... For an example of a transaction shrinker, have a look at:

https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/test/unit/Cardano/Wallet/DBSpec.hs#L276-L282

Reply:
Thanks for all the quickcheck-feedback!